### PR TITLE
Fixed #31486 -- Deprecated passing unsaved model object to related filters.

### PR DESCRIPTION
**Pull request by hramezani**
Fixes [ticket 31486](https://code.djangoproject.com/ticket/31486).

### CODE REVIEW

File: tests/many_to_one/tests.py [hunk 1]
```diff
@@ -626,8 +627,13 @@ def test_relation_unsaved(self):
...
         # execute a query in this case.
-        with self.assertNumQueries(0):
-            self.assertEqual(th.child_set.count(), 0)
+        # TODO: Change to assertRaises after the deprecation period.
+        with self.assertWarnsMessage(
+            RemovedInDjango40Warning,
+            'Passing an unsaved model object to related filter is deprecated.',
+        ):
+            with self.assertNumQueries(0):
+                self.assertEqual(th.child_set.count(), 0)
```
Comment (felixxm) [inline tests/many_to_one/tests.py:636]: IMO this should not raise a warning.

File: django/db/models/fields/related_lookups.py [hunk 2]
```diff
@@ -26,6 +29,15 @@ def get_lookup(self, lookup):
 def get_normalized_value(value, lhs):
     from django.db.models import Model
     if isinstance(value, Model):
+        # When the deprecation ends, replace with:
+        # raise ValueError(
+        #     'The passed object to related filter must be an existing model object.'
```
Comment (felixxm) [inline django/db/models/fields/related_lookups.py:34]: `must be an existing model object.` -> `must be saved.`

File: tests/model_inheritance_regress/tests.py [hunk 3]
```diff
@@ -186,8 +187,13 @@ def test_issue_7488(self):
...
         # case.
-        suppliers = Supplier.objects.filter(
-            restaurant=Restaurant(name='xx', address='yy'))
+        # TODO: Change to assertRaises after the deprecation period.
+        with self.assertWarnsMessage(
+            RemovedInDjango40Warning,
+            'Passing an unsaved model object to related filter is deprecated.',
+        ):
+            suppliers = Supplier.objects.filter(
+                restaurant=Restaurant(name='xx', address='yy'))
```
Comment (felixxm) [inline tests/model_inheritance_regress/tests.py:196]: Have you checked ticket-7488? If admin filters by nonexistent objects then we need to fix this in advance.

File: tests/null_queries/tests.py [hunk 4]
```diff
@@ -42,7 +43,12 @@ def test_none_as_null(self):
 
         # Related managers use __exact=None implicitly if the object hasn't been saved.
         p2 = Poll(question="How?")
-        self.assertEqual(repr(p2.choice_set.all()), '<QuerySet []>')
+        # TODO: Change to assertRaises after the deprecation period.
+        with self.assertWarnsMessage(
+            RemovedInDjango40Warning,
+            'Passing an unsaved model object to related filter is deprecated.',
+        ):
+            self.assertEqual(repr(p2.choice_set.all()), '<QuerySet []>')
```
[the remark this answers is not in the document]
Comment (hramezani) [reply tests/null_queries/tests.py:51]: @felixxm 
Do you have any idea how to fix it?
I don't know how to distinguish between
`p2.choice_set.all()`
and 
`Choice.objects.filter(poll=p2)`

It seems `get_normalized_value` receives same (`value` and `lhs`) for these both cases.

Thanks

[the remark this answers is not in the document]
Comment (felixxm) [reply tests/null_queries/tests.py:51]: Maybe we should move this directly to the `RelatedLookupMixin` :thinking:, I don't have a ready answer.

**Comment by flaiming** [comment]
Can this be reopened? We recently had pretty severe bug because of this. This is really dangerous behavior.

**Comment by felixxm** [comment]
> Can this be reopened? We recently had pretty severe bug because of this. This is really dangerous behavior.

Please feel-free to prepare a new patch if you have an idea how to fix it.

File: tests/model_inheritance_regress/tests.py (hunk 3 above)
Comment (raydeal) [reply tests/model_inheritance_regress/tests.py:196]: The ticket is obsolete and issue in admin doesn't exist.

**Pull request closed without merging**